discoverd/health/register_test.go

```diff
@@ -0,0 +1,255 @@
+			Service: "test",
+			Instance: &discoverd.Instance{
+				Addr: "123",
+			},
+			Registrar: registrar,
+			Monitor:   monitor,
+			Check:     check,
+			Events:    make(chan MonitorEvent),
+		}
+		hb := reg.Register()
+		defer hb.Close()
+
+		errCh := make(chan bool)
+		errCheck := func(ch chan called, stop chan bool) {
+			go func() {
+				select {
+				case _, ok := <-ch:
+					if !ok {
+						return
+					}
+					errCh <- true
+				case <-stop:
+					errCh <- false
+				}
+			}()
+		}
+
+		var stop chan bool
+		for _, step := range steps {
+			stop = make(chan bool)
+			var wait int
+
+			if step.event {
+				monitorChan <- step.up
+				select {
+				case val := <-reg.Events:
+					if step.up {
+						c.Assert(val.Status, Equals, MonitorStatusUp)
+					} else {
+						c.Assert(val.Status, Equals, MonitorStatusDown)
+					}
+				case <-time.After(3 * time.Second):
+					c.Error("Timed out waiting for Registration.Events")
+				}
+			}
+			if step.register {
+				call := <-registrarChan
+				call.returnVal <- step.success
+				<-registrarChan
+			} else {
+				wait++
+				errCheck(registrarChan, stop)
+			}
+			if step.setMeta {
+				go func() {
+					call := <-hbChan
+					call.returnVal <- step.success
+				}()
+				// SetMeta needs to succeed every time, regardless of the situation
+				c.Assert(hb.SetMeta(map[string]string{"TEST": "hello world"}), IsNil)
+			} else {
+				wait++
+				errCheck(hbChan, stop)
+			}
+
+			// check whether functions were called on this step that should not
+			// have ran
+			close(stop)
+			for i := 0; i < wait; i++ {
+				select {
+				case err := <-errCh:
+					if err {
+						c.Error("Received data on a channel that should not be recieving on this step")
+					}
+				}
+			}
+		}
+		close(monitorChan)
+		close(registrarChan)
+		close(hbChan)
+	}
+
+	for _, t := range []struct {
+		name  string
+		steps []step
+	}{
+		{
+			name: "register success up/down/up",
+			steps: []step{
+				{event: true, up: true, register: true, success: true},
+				{event: true},
+				{event: true, up: true, register: true, success: true},
+			},
+		},
+		{
+			name: "register fails then succeeds",
+			steps: []step{
+				{event: true, up: true, register: true, success: false},
+				{register: true, success: true},
+			},
+		},
+		{
+			name: "register is called only once if we get two up events",
+			steps: []step{
+				{event: true, up: true, register: true, success: true},
+				{event: true, up: true, register: false},
+			},
+		},
+		{
+			name: "setmeta while registered",
+			steps: []step{
+				{event: true, up: true, register: true, success: true},
+				{setMeta: true, success: true},
+			},
+		},
+		{
+			name: "setmeta while offline",
+			steps: []step{
+				{setMeta: true, success: false},
+			},
+		},
+		{
+			name: "setmeta while erroring registration",
+			steps: []step{
+				{event: true, up: true, register: true, success: false},
+				{register: true, success: false},
+				{setMeta: true, success: false},
```

Comment:
This and the above check should ensure that when register succeeds the new meta is reflected.